fs/nova/balloc.c

```diff
@@ -569,6 +569,126 @@ static int not_enough_blocks(struct free_list *free_list,
+		(!(num_blocks & PAGES_PER_2MB_MASK) && (atype == DATA))
+
+bool nova_alloc_superpage(struct super_block *sb,
+	struct free_list *free_list, unsigned long num_blocks,
+	unsigned long *new_blocknr, enum nova_alloc_direction from_tail)
+{
+	struct rb_root *tree;
+	struct rb_node *temp;
+	struct nova_range_node *curr;
+	unsigned long curr_blocks;
+	bool found = 0;
+	unsigned long step = 0;
+
+	unsigned int left_margin;
+	unsigned int right_margin;
+
+	tree = &(free_list->block_free_tree);
+	if (from_tail == ALLOC_FROM_HEAD)
+		temp = &(free_list->first_node->node);
+	else
+		temp = &(free_list->last_node->node);
+
+	while (temp) {
+		step++;
+		curr = container_of(temp, struct nova_range_node, node);
+
+		if (!nova_range_node_checksum_ok(curr)) {
+			nova_err(sb, "%s curr failed\n", __func__);
+			goto next;
+		}
+
+		curr_blocks = curr->range_high - curr->range_low + 1;
+		left_margin = PAGES_PER_2MB - (curr->range_low & PAGES_PER_2MB_MASK);
+
+		/*
+		 * Guard against cases where:
+		 * a. Unaligned free blocks is smaller than #512
+		 *    left_margin could larger than curr_blocks.
+		 * b. After alignment, free blocks is smaller than
+		 *    requested blocks.
+		 * Otherwise, we are free to go.
+		 */
+		if ((curr_blocks > left_margin) && \
+			(num_blocks <= (curr_blocks - left_margin))) {
+			struct nova_range_node *node;
+			unsigned long saved_range_high;
+
+			*new_blocknr = curr->range_low + left_margin;
+			right_margin = curr_blocks - left_margin - num_blocks;
+			nova_dbgv("curr:%p: num_blocks:%lu curr->range_low:%lu high:%lu",
+						curr, num_blocks, curr->range_low, curr->range_high);
+
+			if (left_margin) {
+				/* Reuse "curr" and its "first_node" indicator. */
+				saved_range_high = curr->range_high;
+				curr->range_high = curr->range_low + left_margin - 1;
+				nova_update_range_node_checksum(curr);
+				nova_dbgv("Insert node for left_margin, range_low:%lu high:%lu",
+							curr->range_low, curr->range_high);
+			}
+
+			if (right_margin) {
+				if (left_margin) {
+					/* curr was reused for left_margin node, grab new one. */
+					node = nova_alloc_blocknode_atomic(sb);
+					if (node == NULL) {
+						nova_warn("Failed to allocate new block node.\n");
+						return -ENOMEM;
+					}
+					node->range_low = curr->range_low + left_margin + num_blocks;
+					node->range_high = saved_range_high;
+					nova_update_range_node_checksum(node);
+					nova_insert_blocktree(tree, node);
+					free_list->num_blocknode++;
+					if (curr == free_list->last_node)
+						free_list->last_node = node;
+				} else {
+					/*
+					 * curr->range_low is aligned, reuse curr for right_margin.
+					 * RB tree is arragned by node->range_low as key, so once
+					 * range_low is changed, erase and reinsert is required.
+					 */
+					rb_erase(&curr->node, tree);
+					curr->range_low = curr->range_low + num_blocks;
+					nova_update_range_node_checksum(curr);
+					nova_insert_blocktree(tree, curr);
```

Comment:
I don't think you need to erase - insert here.

Reply:
So how does this look?

                    curr->range_low = curr->range_low + num_blocks;
                    nova_update_range_node_checksum(curr);

Reply:
Yes.